sled-agent/config-reconciler/src/reconciler_task/external_disks.rs

```diff
@@ -0,0 +1,1020 @@
+        // If we made any changes, update the set of disks visbile to external
+        // consumers. (It would be correct to call this unconditionally, but we
+        // can save a bit of work by skipping it in the common case of "no disks
+        // were removed".)
+        if !disk_ids_to_remove.is_empty() || marked_disk_not_found {
+            self.update_currently_managed_zpools();
+        }
+    }
+
+    /// Attempt to start managing any disks specified by `config` that we aren't
+    /// already managing.
+    pub(super) async fn start_managing_if_needed(
+        &mut self,
+        raw_disks: &IdMap<RawDiskWithId>,
+        config: &IdMap<OmicronPhysicalDiskConfig>,
+        key_requester: &StorageKeyRequester,
+        log: &Logger,
+    ) {
+        self.start_managing_if_needed_with_disk_adopter(
+            raw_disks,
+            config,
+            log,
+            &mut RealDiskAdopter { key_requester },
+        )
+        .await
+    }
+
+    async fn start_managing_if_needed_with_disk_adopter<T: DiskAdopter>(
+        &mut self,
+        raw_disks: &IdMap<RawDiskWithId>,
+        config: &IdMap<OmicronPhysicalDiskConfig>,
+        log: &Logger,
+        disk_adopter: &mut T,
+    ) {
+        for config in config.iter().cloned() {
+            // We can only manage disks if the raw disk is present.
+            let Some(raw_disk) = raw_disks.get(&config.identity) else {
+                warn!(
+                    log,
+                    "Control plane disk requested, but not detected within sled";
+                    "disk_identity" => ?&config.identity
+                );
+                let err = DiskManagementError::NotFound;
+                self.disks.insert(ExternalDiskState::failed(config, err));
+                continue;
+            };
+
+            // Refuse to manage internal disks.
+            match raw_disk.variant() {
+                DiskVariant::U2 => (),
+                DiskVariant::M2 => {
+                    warn!(
+                        log,
+                        "Control plane requested management of internal disk";
+                        "config" => ?config,
+                    );
+                    let err =
+                        DiskManagementError::InternalDiskControlPlaneRequest(
+                            config.id,
+                        );
+                    self.disks.insert(ExternalDiskState::failed(config, err));
+                    continue;
+                }
+            }
+
+            // Take ownership of our current disk state (if we have one), then
+            // immediately put back the new state.
+            let maybe_current = self.disks.remove(&config.id);
+            self.disks.insert(
+                self.try_ensure_disk_managed(
+                    maybe_current,
+                    config,
+                    raw_disk,
+                    disk_adopter,
+                    log,
+                )
+                .await,
+            );
```

Comment:
Given that "disk adoption" might involve creating partitions on the underlying disks, it seems like it might be nice to do this operation concurrently?

This appears to currently be:

- Take out the Option<State> for the disk
- Call "try_ensure_disk_managed" to transform that state (or create it, if it was None)
- Put that state back into `self.disks`

Each operation seems to be acting on independent parts of `self.disks`, and I don't *think* the `disk_adopter` actually needs mutability. WDYT about doing this concurrently?

Reply:
(I care less about this in the removal path, because that's significantly cheaper)